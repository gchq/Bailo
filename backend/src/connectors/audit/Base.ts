import { Request } from 'express'

import { AccessRequestDoc } from '../../models/AccessRequest.js'
import { FileInterface, FileInterfaceDoc } from '../../models/File.js'
import { InferenceDoc } from '../../models/Inference.js'
import { ModelCardInterface, ModelDoc, ModelInterface } from '../../models/Model.js'
import { ReleaseDoc } from '../../models/Release.js'
import { ResponseInterface } from '../../models/Response.js'
import { ReviewInterface } from '../../models/Review.js'
import { SchemaDoc, SchemaInterface } from '../../models/Schema.js'
import { TokenDoc } from '../../models/Token.js'
import { ModelSearchResult } from '../../routes/v2/model/getModelsSearch.js'
import { BailoError } from '../../types/error.js'

const AuditKind = {
  Create: 'Create',
  View: 'View',
  Update: 'Update',
  Delete: 'Delete',
  Search: 'Search',
} as const
export type AuditKindKeys = (typeof AuditKind)[keyof typeof AuditKind]

export const AuditInfo = {
  CreateModel: { typeId: 'CreateModel', description: 'Model Created', auditKind: AuditKind.Create },
  ViewModel: { typeId: 'ViewModel', description: 'Model Viewed', auditKind: AuditKind.View },
  UpdateModel: { typeId: 'UpdateModel', description: 'Model Updated', auditKind: AuditKind.Update },
  SearchModels: { typeId: 'SearchModels', description: 'Model Searched', auditKind: AuditKind.Search },

  CreateModelCard: { typeId: 'CreateModelCard', description: 'Model Card Created', auditKind: AuditKind.Create },
  ViewModelCard: { typeId: 'ViewModelCard', description: 'Model Card Viewed', auditKind: AuditKind.View },
  ViewModelCardRevisions: {
    typeId: 'ViewModelCardRevisions',
    description: 'Model Card Revisions Viewed',
    auditKind: AuditKind.View,
  },
  UpdateModelCard: { typeId: 'UpdateModelCard', description: 'Model Card Updated', auditKind: AuditKind.Update },

  CreateFile: { typeId: 'CreateFile', description: 'File Information Created', auditKind: AuditKind.Create },
  ViewFile: { typeId: 'ViewFile', description: 'File Downloaded', auditKind: AuditKind.View },
  ViewFiles: { typeId: 'ViewFiles', description: 'File Information Viewed', auditKind: AuditKind.View },
  DeleteFile: { typeId: 'DeleteFile', description: 'File Information Deleted', auditKind: AuditKind.Delete },

  CreateRelease: { typeId: 'CreateRelease', description: 'Release Created', auditKind: AuditKind.Create },
  ViewRelease: { typeId: 'ViewRelease', description: 'Release Viewed', auditKind: AuditKind.View },
  UpdateRelease: { typeId: 'UpdateRelease', description: 'Release Updated', auditKind: AuditKind.Update },
  DeleteRelease: { typeId: 'DeleteRelease', description: 'Release Deleted', auditKind: AuditKind.Delete },
  ViewReleases: { typeId: 'ViewReleases', description: 'Releases Viewed', auditKind: AuditKind.View },

  CreateUserToken: { typeId: 'CreateUserToken', description: 'Token Created', auditKind: AuditKind.Create },
  ViewUserTokens: { typeId: 'ViewUserToken', description: 'Token Viewed', auditKind: AuditKind.View },
  DeleteUserToken: { typeId: 'DeleteUserToken', description: 'Token Deleted', auditKind: AuditKind.Delete },

  CreateAccessRequest: {
    typeId: 'CreateAccessRequest',
    description: 'Access Request Created',
    auditKind: AuditKind.Create,
  },
  ViewAccessRequest: { typeId: 'ViewAccessRequest', description: 'Access Request Viewed', auditKind: AuditKind.View },
  UpdateAccessRequest: {
    typeId: 'UpdateAccess Request',
    description: 'Access Request Updated',
    auditKind: AuditKind.Update,
  },
  DeleteAccessRequest: {
    typeId: 'UpdateAccessRequest',
    description: 'Access Request Deleted',
    auditKind: AuditKind.Update,
  },
  ViewAccessRequests: {
    typeId: 'ViewAccessRequests',
    description: 'Access Requests Viewed',
    auditKind: AuditKind.View,
  },

  SearchReviews: {
    typeId: 'SearchReviews',
    description: 'Reviews Searched',
    auditKind: AuditKind.Search,
  },
  CreateReviewResponse: {
    typeId: 'CreateReviewResponse',
    description: 'Review Response Created',
    auditKind: AuditKind.Create,
  },
  UpdateReviewResponse: {
    typeId: 'UpdateReviewResponse',
    description: 'Review Response Updated',
    auditKind: AuditKind.Update,
  },

  CreateSchema: { typeId: 'CreateSchema', description: 'Schema Created', auditKind: AuditKind.Create },
  SearchSchemas: { typeId: 'SearchedSchemas', description: 'Schemas Searched', auditKind: AuditKind.Search },
  ViewSchema: { typeId: 'ViewSchema', description: 'Schema Viewed', auditKind: AuditKind.View },
  DeleteSchema: { typeId: 'DeleteSchema', description: 'Schema Deleted', auditKind: AuditKind.Delete },
  UpdateSchema: { typeId: 'UpdateSchema', description: 'Schema Updated', auditKind: AuditKind.Update },

  ViewModelImages: { typeId: 'ViewModelImages', description: 'Model Images Viewed', auditKind: AuditKind.View },

  CreateInference: { typeId: 'CreateInference', description: 'Inference Service Created', auditKind: AuditKind.Create },
  UpdateInference: { typeId: 'UpdateInference', description: 'Inference Service Updated', auditKind: AuditKind.Update },
  ViewInference: { typeId: 'ViewInference', description: 'Inference Service Viewed', auditKind: AuditKind.View },
  ViewInferences: { typeId: 'ViewInferences', description: 'Inferences Viewed', auditKind: AuditKind.View },

  CreateExport: { typeId: 'CreateExport', description: 'Model Exported', auditKind: AuditKind.Create },
  CreateImport: { typeId: 'CreateImport', description: 'Model Imported', auditKind: AuditKind.Create },

  ViewResponses: {
    typeId: 'ViewResponses',
    description: 'View a list of responses',
    auditKind: AuditKind.View,
  },
  CreateResponse: {
    typeId: 'CreateResponse',
    description: 'Review or comment responses created',
    auditKind: AuditKind.Create,
  },
  UpdateResponse: {
    typeId: 'UpdateResponse',
    description: 'Updated a comment or review response',
    auditKind: AuditKind.Update,
  },
} as const
export type AuditInfoKeys = (typeof AuditInfo)[keyof typeof AuditInfo]

export abstract class BaseAuditConnector {
  abstract onCreateModel(req: Request, model: ModelDoc)
  abstract onViewModel(req: Request, model: ModelDoc)
  abstract onUpdateModel(req: Request, model: ModelDoc)
  abstract onSearchModel(req: Request, models: ModelSearchResult[])

  abstract onCreateModelCard(req: Request, modelId: string, modelCard: ModelCardInterface)
  abstract onViewModelCard(req: Request, modelId: string, modelCard: ModelCardInterface)
  abstract onUpdateModelCard(req: Request, modelId: string, modelCard: ModelCardInterface)
  abstract onViewModelCardRevisions(req: Request, modelId: string, modelCards: ModelCardInterface[])

  abstract onCreateFile(req: Request, file: FileInterfaceDoc)
  abstract onViewFile(req: Request, file: FileInterfaceDoc)
  abstract onViewFiles(req: Request, modelId: string, files: FileInterface[])
  abstract onDeleteFile(req: Request, modelId: string, fileId: string)

  abstract onCreateRelease(req: Request, release: ReleaseDoc)
  abstract onViewRelease(req: Request, release: ReleaseDoc)
  abstract onUpdateRelease(req: Request, release: ReleaseDoc)
  abstract onDeleteRelease(req: Request, modelId: string, semver: string)
  abstract onViewReleases(req: Request, releases: ReleaseDoc[])

  abstract onCreateCommentResponse(req: Request, response: ResponseInterface)
  abstract onCreateReviewResponse(req: Request, response: ResponseInterface)

  abstract onViewResponses(req: Request, responseInterfaces: ResponseInterface[])
  abstract onUpdateResponse(req: Request, responseId: string)

  abstract onCreateUserToken(req: Request, token: TokenDoc)
  abstract onViewUserTokens(req: Request, tokens: TokenDoc[])
  abstract onDeleteUserToken(req: Request, accessKey: string)

  abstract onCreateAccessRequest(req: Request, accessRequest: AccessRequestDoc)
  abstract onViewAccessRequest(req: Request, accessRequest: AccessRequestDoc)
  abstract onUpdateAccessRequest(req: Request, accessRequest: AccessRequestDoc)
  abstract onDeleteAccessRequest(req: Request, accessRequestId: string)
  abstract onViewAccessRequests(req: Request, accessRequests: AccessRequestDoc[])

  abstract onSearchReviews(req: Request, reviews: (ReviewInterface & { model: ModelInterface })[])

  abstract onSearchSchemas(req: Request, schemas: SchemaInterface[])
  abstract onCreateSchema(req: Request, schema: SchemaInterface)
  abstract onViewSchema(req: Request, schema: SchemaInterface)
  abstract onDeleteSchema(req: Request, schemaId: string)
  abstract onUpdateSchema(req: Request, schema: SchemaDoc)

  abstract onCreateInference(req: Request, inference: InferenceDoc)
  abstract onUpdateInference(req: Request, inference: InferenceDoc)
  abstract onViewInference(req: Request, inference: InferenceDoc)
  abstract onViewInferences(req: Request, inference: InferenceDoc[])

  abstract onViewModelImages(
    req: Request,
    modelId: string,
    images: { repository: string; name: string; tags: string[] }[],
  )

  abstract onCreateS3Export(req: Request, modelId: string, semvers?: string[])
  abstract onCreateImport(
    req: Request,
    mirroredModelId: string,
    sourceModelId: string,
    modelCardVersions: number[],
    exporter: string,
  )

  abstract onError(req: Request, error: BailoError)

  checkEventType(auditInfo: AuditInfoKeys, req: Request) {
    if (auditInfo.typeId !== req.audit.typeId && auditInfo.description !== req.audit.description) {
      throw new Error(`Audit: Expected type '${JSON.stringify(auditInfo)}' but recieved '${JSON.stringify(req.audit)}'`)
    }
  }
}
