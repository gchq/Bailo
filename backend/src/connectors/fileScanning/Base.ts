import { FileInterface } from '../../models/File.js'
import { ScanInterface } from '../../models/Scan.js'
import log from '../../services/log.js'
export type FileScanResult = Pick<
  ScanInterface,
  'toolName' | 'scannerVersion' | 'state' | 'isInfected' | 'viruses' | 'lastRunAt'
>

export const ScanState = {
  NotScanned: 'notScanned',
  InProgress: 'inProgress',
  Complete: 'complete',
  Error: 'error',
} as const
export type ScanStateKeys = (typeof ScanState)[keyof typeof ScanState]

export type FileScanningConnectorInfo = Pick<FileScanResult, 'toolName' | 'scannerVersion'>

export abstract class BaseFileScanningConnector {
  abstract toolName: string
  abstract version: string | undefined

  abstract init()
  abstract scan(file: FileInterface): Promise<FileScanResult[]>

  info(): FileScanningConnectorInfo {
    return { toolName: this.toolName, scannerVersion: this.version }
  }

  async scanError(message: string, context?: object): Promise<FileScanResult[]> {
    const scannerInfo = await this.info()
    log.error({ ...context, ...scannerInfo }, message)
    return [
      {
        ...scannerInfo,
        state: ScanState.Error,
        lastRunAt: new Date(),
      },
    ]
  }
}
