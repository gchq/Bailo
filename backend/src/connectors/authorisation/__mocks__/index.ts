import { vi } from 'vitest'

const auth = {
  hasModelVisibilityAccess: vi.fn(() => true),
  hasAccessRequest: vi.fn(() => true),

  model: vi.fn(() => ({ success: true })),
  models: vi.fn(() => [{ success: true }]),

  schema: vi.fn(() => ({ success: true })),
  schemas: vi.fn(() => [{ success: true }]),

  schemaMigration: vi.fn(() => ({ success: true })),
  schemaMigrations: vi.fn(() => ({ success: true })),

  accessRequest: vi.fn(() => ({ success: true })),
  accessRequests: vi.fn(() => [{ success: true }]),

  release: vi.fn(() => ({ success: true })),
  releases: vi.fn(() => [{ success: true }]),

  file: vi.fn(() => ({ success: true })),
  files: vi.fn(() => [{ success: true }]),

  image: vi.fn(() => ({ success: true })),
  images: vi.fn(() => [{ success: true }]),
}

export default auth
