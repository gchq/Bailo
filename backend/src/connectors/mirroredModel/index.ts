import PQueue from 'p-queue'

import config from '../../utils/config.js'
import { DocumentsMirrorMetadata, MongoDocumentMirrorInformation } from './documents.js'
import { FileMirrorInformation, FileMirrorMetadata } from './file.js'
import { ImageMirrorInformation, ImageMirrorMetadata } from './image.js'

export const MirrorKind = {
  Documents: 'documents',
  File: 'file',
  Image: 'image',
} as const

export type MirrorKindKeys<T extends keyof typeof MirrorKind | void = void> = T extends keyof typeof MirrorKind
  ? (typeof MirrorKind)[T]
  : (typeof MirrorKind)[keyof typeof MirrorKind]

export type MirrorMetadata = DocumentsMirrorMetadata | FileMirrorMetadata | ImageMirrorMetadata
export type MirrorInformation = MongoDocumentMirrorInformation | FileMirrorInformation | ImageMirrorInformation

export const exportQueue: PQueue = new PQueue({ concurrency: config.modelMirror.export.concurrency })
