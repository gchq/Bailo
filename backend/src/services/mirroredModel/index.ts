import { DocumentsMirrorMetadata, MongoDocumentMirrorInformation } from './importers/documents.js'
import { FileMirrorInformation, FileMirrorMetadata } from './importers/file.js'
import { ImageMirrorInformation, ImageMirrorMetadata } from './importers/image.js'

export const MirrorKind = {
  Documents: 'documents',
  File: 'file',
  Image: 'image',
} as const

export type MirrorKindKeys<T extends keyof typeof MirrorKind | void = void> = T extends keyof typeof MirrorKind
  ? (typeof MirrorKind)[T]
  : (typeof MirrorKind)[keyof typeof MirrorKind]

export type MirrorMetadata = DocumentsMirrorMetadata | FileMirrorMetadata | ImageMirrorMetadata
export type MirrorInformation = MongoDocumentMirrorInformation | FileMirrorInformation | ImageMirrorInformation
