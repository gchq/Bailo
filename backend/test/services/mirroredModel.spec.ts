import { PassThrough } from 'stream'
import { describe, expect, test, vi } from 'vitest'

import authorisation from '../../src/connectors/authorisation/index.js'
import { FileScanResult } from '../../src/connectors/fileScanning/Base.js'
import { UserInterface } from '../../src/models/User.js'
import { exportModel, importModel } from '../../src/services/mirroredModel.js'

const fileScanResult: FileScanResult = {
  state: 'complete',
  isInfected: false,
  toolName: 'Test',
}

const fileScanningMock = vi.hoisted(() => ({
  info: vi.fn(() => []),
  scan: vi.fn(() => new Promise(() => [fileScanResult])),
}))
vi.mock('../../src/connectors/fileScanning/index.js', async () => ({ default: fileScanningMock }))

const fflateMock = vi.hoisted(() => ({
  unzipSync: vi.fn(),
}))
vi.mock('fflate', async () => fflateMock)

const bufferMock = vi.hoisted(() => ({
  unzipSync: vi.fn(),
}))
vi.mock('node:buffer', async () => bufferMock)

const fetchMock = vi.hoisted(() => ({
  default: vi.fn(() => ({ ok: true, body: vi.fn(), text: vi.fn() })),
}))
vi.mock('node-fetch', async () => fetchMock)

const authMock = vi.hoisted(() => ({
  model: vi.fn(() => ({ success: true })),
}))
vi.mock('../../src/connectors/authorisation/index.js', async () => ({
  default: authMock,
}))

const configMock = vi.hoisted(
  () =>
    ({
      ui: {
        modelMirror: {
          import: {
            enabled: true,
          },
          export: {
            enabled: true,
          },
        },
      },
      s3: { buckets: { uploads: 'test' } },
      modelMirror: {
        export: {
          maxSize: 100,
          kmsSignature: {
            enabled: true,
          },
        },
      },
    }) as any,
)
vi.mock('../../src/utils/config.js', () => ({
  __esModule: true,
  default: configMock,
}))

const logMock = vi.hoisted(() => ({
  info: vi.fn(),
  debug: vi.fn(),
  warn: vi.fn(),
  error: vi.fn(),
}))
vi.mock('../../src/services/log.js', async () => ({
  default: logMock,
}))

const modelMocks = vi.hoisted(() => ({
  getModelById: vi.fn(() => ({ settings: { mirror: { destinationModelId: '123' } } })),
  getModelCardRevisions: vi.fn(() => [{ toJSON: vi.fn(), version: 123 }]),
  setLatestImportedModelCard: vi.fn(),
  saveImportedModelCard: vi.fn(),
  isModelCardRevision: vi.fn(() => true),
}))
vi.mock('../../src/services/model.js', () => modelMocks)

const releaseMocks = vi.hoisted(() => ({
  getReleasesForExport: vi.fn(() => [{ toJSON: vi.fn() }]),
  getAllFileIds: vi.fn(() => [{}]),
}))
vi.mock('../../src/services/release.js', () => releaseMocks)

const fileMocks = vi.hoisted(() => ({
  getFilesByIds: vi.fn(() => [{ _id: '123', avScan: [{ state: 'complete', isInfected: false }], toJSON: vi.fn() }]),
  getTotalFileSize: vi.fn(() => 42),
  downloadFile: vi.fn(() => ({ Body: 'test' })),
}))
vi.mock('../../src/services/file.js', () => fileMocks)

const archiverMocks = vi.hoisted(() => ({
  append: vi.fn(),
  finalize: vi.fn(),
  pipe: vi.fn(),
}))
vi.mock('archiver', () => ({ default: vi.fn(() => archiverMocks) }))

const s3Mocks = vi.hoisted(() => ({
  putObjectStream: vi.fn(() => Promise.resolve({ fileSize: 100 })),
  getObjectStream: vi.fn(() => Promise.resolve({ Body: new PassThrough() })),
}))
vi.mock('../../src/clients/s3.js', () => s3Mocks)

const kmsMocks = vi.hoisted(() => ({
  sign: vi.fn(),
}))
vi.mock('../../src/clients/kms.js', () => kmsMocks)

const hashMocks = vi.hoisted(() => ({
  createHash: vi.fn(() => ({
    setEncoding: vi.fn(),
    end: vi.fn(),
    read: vi.fn(() => 'test digest'),
  })),
}))
vi.mock('node:crypto', () => hashMocks)

const streamMocks = vi.hoisted(() => ({
  PassThrough: vi.fn(() => ({
    pipe: vi.fn(),
    on: vi.fn((a, b) => {
      b()
    }),
  })),
}))
vi.mock('stream', () => streamMocks)

describe('services > mirroredModel', () => {
  test('exportModel > not enabled', async () => {
    vi.spyOn(configMock, 'ui', 'get').mockReturnValueOnce({ modelMirror: { export: { enabled: false } } })
    const response = exportModel({} as UserInterface, 'modelId', true)

    await expect(response).rejects.toThrowError('Exporting models has not been enabled.')
  })

  test('exportModel > bad authorisation', async () => {
    vi.mocked(authorisation.model).mockResolvedValueOnce({ info: 'You do not have permission', success: false, id: '' })

    const response = exportModel({} as UserInterface, 'modelId', true)
    await expect(response).rejects.toThrowError(/^You do not have permission/)
  })

  test('exportModel > missing disclaimer agreement', async () => {
    const response = exportModel({} as UserInterface, 'modelId', false)
    await expect(response).rejects.toThrowError(
      /^You must agree to the disclaimer agreement before being able to export a model./,
    )
  })

  test('exportModel > unable to create model card zip file', async () => {
    archiverMocks.append.mockImplementationOnce(() => {
      throw Error('Error making zip file')
    })
    const response = exportModel({} as UserInterface, 'modelId', true)
    await expect(response).rejects.toThrowError(/^Error when adding the model card revision\(s\) to the zip file./)
  })

  test('exportModel > unable to create release zip file', async () => {
    archiverMocks.append.mockReturnValueOnce({})
    archiverMocks.append.mockImplementationOnce(() => {
      throw Error('Error making zip file')
    })
    const response = exportModel({} as UserInterface, 'modelId', true, ['1.2.3'])
    await expect(response).rejects.toThrowError(/^Error when adding the release\(s\) to the zip file./)
  })

  test('exportModel > unable to create digest for zip file', async () => {
    hashMocks.createHash.mockImplementationOnce(() => {
      throw Error()
    })
    await exportModel({} as UserInterface, 'modelId', true, ['1.2.3'])
    await new Promise((r) => setTimeout(r))
    await expect(logMock.error).toBeCalledWith(
      expect.any(Object),
      'Failed to upload export to export location with signatures',
    )
  })

  test('exportModel > unable to create kms signature for zip file', async () => {
    kmsMocks.sign.mockRejectedValueOnce('Error')
    await exportModel({} as UserInterface, 'modelId', true, ['1.2.3'])
    await new Promise((r) => setTimeout(r))
    await expect(logMock.error).toBeCalledWith(
      expect.any(Object),
      'Failed to upload export to export location with signatures',
    )
  })

  test('exportModel > release export size too large', async () => {
    vi.spyOn(configMock, 'modelMirror', 'get').mockReturnValue({
      enabled: true,
      export: {
        maxSize: 10,
        kmsSignature: {
          enabled: true,
        },
      },
    })
    fileMocks.getTotalFileSize.mockReturnValueOnce(100)
    const response = exportModel({} as UserInterface, 'modelId', true, ['1.2.3', '1.2.4'])
    await expect(response).rejects.toThrowError(/^Requested export is too large./)
  })

  test('exportModel > successful export if no files exist', async () => {
    releaseMocks.getAllFileIds.mockResolvedValueOnce([])
    await exportModel({} as UserInterface, 'modelId', true, ['1.2.3', '1.2.4'])
    // Allow for completion of asynchronous content
    await new Promise((r) => setTimeout(r))
    await expect(s3Mocks.putObjectStream).toBeCalledTimes(2)
  })

  test('exportModel > missing mirrored model ID', async () => {
    modelMocks.getModelById.mockReturnValueOnce({
      settings: { mirror: { destinationModelId: '' } },
    })
    const response = exportModel({} as UserInterface, 'modelId', true, ['1.2.3'])
    await expect(response).rejects.toThrowError(/^The 'Destination Model ID' has not been set on this model./)
    await expect(s3Mocks.putObjectStream).toHaveBeenCalledTimes(0)
  })

  test('exportModel > export contains infected file', async () => {
    fileMocks.getFilesByIds.mockReturnValueOnce([
      { _id: '123', avScan: [{ state: 'complete', isInfected: true }], toJSON: vi.fn() },
      { _id: '321', avScan: [{ state: 'complete', isInfected: false }], toJSON: vi.fn() },
    ])
    const response = exportModel({} as UserInterface, 'modelId', true, ['1.2.3'])
    await expect(response).rejects.toThrowError('The releases contain file(s) that do not have a clean AV scan.')
    await expect(s3Mocks.putObjectStream).toHaveBeenCalledTimes(0)
  })

  test('exportModel > export contains incomplete file scan', async () => {
    fileMocks.getFilesByIds.mockReturnValueOnce([
      { _id: '123', avScan: [{ state: 'inProgress' }] as any, toJSON: vi.fn() },
      { _id: '321', avScan: [{ state: 'complete', isInfected: false }], toJSON: vi.fn() },
    ])
    const response = exportModel({} as UserInterface, 'modelId', true, ['1.2.3'])
    await expect(response).rejects.toThrowError('The releases contain file(s) that do not have a clean AV scan.')
    await expect(s3Mocks.putObjectStream).toHaveBeenCalledTimes(0)
  })

  test('exportModel > export missing file scan', async () => {
    fileMocks.getFilesByIds.mockReturnValueOnce([
      { _id: '123', toJSON: vi.fn() } as any,
      { _id: '321', avScan: [{ state: 'complete', isInfected: false }], toJSON: vi.fn() },
      { _id: '321', avScan: [{ state: 'complete', isInfected: false }], toJSON: vi.fn() },
    ])
    const response = exportModel({} as UserInterface, 'testmod', true, ['1.2.3'])
    await expect(response).rejects.toThrowError('The releases contain file(s) that do not have a clean AV scan.')
    await expect(s3Mocks.putObjectStream).toHaveBeenCalledTimes(0)
  })

  test('exportModel > upload straight to the export bucket if signatures are disabled', async () => {
    vi.spyOn(configMock, 'modelMirror', 'get').mockReturnValue({
      enabled: true,
      export: {
        bucket: 'exports',
        kmsSignature: {
          enabled: false,
        },
      },
    })

    await exportModel({} as UserInterface, 'modelId', true)
    expect(s3Mocks.putObjectStream).toHaveBeenCalledWith(
      'exports',
      'modelId.zip',
      expect.any(Object),
      expect.any(Object),
    )
    expect(s3Mocks.putObjectStream).toHaveBeenCalledTimes(1)
  })

  test('exportModel > log error if unable to upload to export s3 storage', async () => {
    s3Mocks.putObjectStream.mockRejectedValueOnce('')
    vi.spyOn(configMock, 'modelMirror', 'get').mockReturnValue({
      enabled: true,
      export: {
        kmsSignature: {
          enabled: false,
        },
      },
    })

    await exportModel({} as UserInterface, 'modelId', true)
    expect(logMock.error).toHaveBeenCalledWith(expect.any(Object), 'Failed to export to export S3 location.')
  })

  test('exportModel > export uploaded to S3 with signatures', async () => {
    vi.spyOn(configMock, 'modelMirror', 'get').mockReturnValue({
      enabled: true,
      export: {
        kmsSignature: {
          enabled: true,
        },
      },
    })
    await exportModel({} as UserInterface, 'modelId', true)
    expect(s3Mocks.putObjectStream).toBeCalledTimes(1)
  })

  test('exportModel > export uploaded to S3 for model cards and releases', async () => {
    await exportModel({} as UserInterface, 'modelId', true, ['1.2.3', '3.2.1'])

    expect(s3Mocks.putObjectStream).toBeCalledTimes(2)
  })

  test('exportModel > unable to upload to tmp S3 location', async () => {
    s3Mocks.putObjectStream.mockRejectedValueOnce('')
    await exportModel({} as UserInterface, 'modelId', true, ['1.2.3'])
    await new Promise((r) => setTimeout(r))
    expect(logMock.error).toBeCalledWith(expect.any(Object), 'Failed to export to temporary S3 location.')
  })

  test('exportModel > unable to get object from tmp S3 location', async () => {
    s3Mocks.getObjectStream.mockRejectedValueOnce('')
    await exportModel({} as UserInterface, 'modelId', true, ['1.2.3'])
    await new Promise((r) => setTimeout(r))
    expect(logMock.error).toBeCalledWith(expect.any(Object), 'Failed to retrieve stream from temporary S3 location.')
  })

  test('importModel > not enabled', async () => {
    vi.spyOn(configMock, 'ui', 'get').mockReturnValueOnce({ modelMirror: { import: { enabled: false } } })
    const result = importModel({} as UserInterface, '', 'https://test.com')

    await expect(result).rejects.toThrowError('Importing models has not been enabled.')
  })

  test('importModel > mirrored model Id empty', async () => {
    const result = importModel({} as UserInterface, '', 'https://test.com')

    await expect(result).rejects.toThrowError('Missing mirrored model ID.')
  })

  test('importModel > error when getting zip file', async () => {
    fetchMock.default.mockRejectedValueOnce('a')
    const result = importModel({} as UserInterface, 'model-id', 'https://test.com')

    await expect(result).rejects.toThrowError('Unable to get the file.')
  })

  test('importModel > non 200 response when getting zip file', async () => {
    fetchMock.default.mockResolvedValueOnce({ ok: false, body: vi.fn(), text: vi.fn() })
    const result = importModel({} as UserInterface, 'model-id', 'https://test.com')

    await expect(result).rejects.toThrowError('Unable to get zip file.')
  })

  test('importModel > file missing from body', async () => {
    fetchMock.default.mockResolvedValueOnce({ ok: true, text: vi.fn() } as any)
    const result = importModel({} as UserInterface, 'model-id', 'https://test.com')

    await expect(result).rejects.toThrowError('Unable to get the file.')
  })

  test('importModel > save each imported model card', async () => {
    fetchMock.default.mockResolvedValueOnce({ ok: true, body: vi.fn(), text: vi.fn(), arrayBuffer: vi.fn() } as any)
    fflateMock.unzipSync.mockReturnValueOnce({
      file1: Buffer.from(JSON.stringify({ modelId: 'abc' })),
      file2: Buffer.from(JSON.stringify({ modelId: 'abc' })),
    })
    await importModel({} as UserInterface, 'model-id', 'https://test.com')

    await expect(modelMocks.saveImportedModelCard.mock.calls.length).toBe(2)
  })

  test('importModel > cannot parse into model card', async () => {
    fetchMock.default.mockResolvedValueOnce({ ok: true, body: vi.fn(), text: vi.fn(), arrayBuffer: vi.fn() } as any)
    fflateMock.unzipSync.mockReturnValueOnce({
      file1: Buffer.from(JSON.stringify({})),
    })
    modelMocks.isModelCardRevision.mockReturnValueOnce(false)
    const result = importModel({} as UserInterface, 'model-id', 'https://test.com')

    await expect(result).rejects.toThrowError(/^Data cannot be converted into a model card./)
  })

  test('importModel > different model IDs in zip files', async () => {
    fetchMock.default.mockResolvedValueOnce({ ok: true, body: vi.fn(), text: vi.fn(), arrayBuffer: vi.fn() } as any)
    fflateMock.unzipSync.mockReturnValueOnce({
      file1: Buffer.from(JSON.stringify({ modelId: 'abc' })),
      file2: Buffer.from(JSON.stringify({ modelId: 'cba' })),
    })
    const result = importModel({} as UserInterface, 'model-id', 'https://test.com')

    await expect(result).rejects.toThrowError(/^Zip file contains model cards for multiple models./)
  })

  test('importModel > invalid zip data', async () => {
    fetchMock.default.mockResolvedValueOnce({ ok: true, body: vi.fn(), text: vi.fn(), arrayBuffer: vi.fn() } as any)
    fflateMock.unzipSync.mockImplementationOnce(() => {
      throw Error('Cannot import file.')
    })
    const result = importModel({} as UserInterface, 'model-id', 'https://test.com')

    await expect(result).rejects.toThrowError(/^Unable to read zip file./)
  })
})
