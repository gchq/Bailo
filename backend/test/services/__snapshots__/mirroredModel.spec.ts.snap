// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`services > mirroredModel > importModel > save each imported file 1`] = `
{
  "importResult": {
    "fileIds": [
      undefined,
      undefined,
    ],
    "imageIds": [],
    "modelCardVersions": [],
    "newModelCards": [],
    "newReleases": [],
    "releaseSemvers": [],
  },
  "mirroredModel": undefined,
}
`;

exports[`services > mirroredModel > importModel > save each imported model card 1`] = `
{
  "importResult": {
    "fileIds": [],
    "imageIds": [],
    "modelCardVersions": [
      undefined,
      undefined,
    ],
    "newModelCards": [],
    "newReleases": [],
    "releaseSemvers": [],
  },
  "mirroredModel": undefined,
}
`;

exports[`services > mirroredModel > importModel > save each imported release 1`] = `
{
  "importResult": {
    "fileIds": [],
    "imageIds": [
      "localhost:8080/imageName:tag",
    ],
    "modelCardVersions": [],
    "newModelCards": [],
    "newReleases": [
      {
        "modelId": "source-model-id",
      },
      {
        "modelId": "source-model-id",
      },
    ],
    "releaseSemvers": [
      undefined,
      undefined,
    ],
  },
  "mirroredModel": undefined,
}
`;

exports[`services > mirroredModel > importModel > skip non-File entry 1`] = `
{
  "importResult": {
    "fileIds": [],
    "imageIds": [],
    "modelCardVersions": [],
    "newModelCards": [],
    "newReleases": [],
    "releaseSemvers": [],
  },
  "mirroredModel": undefined,
}
`;

exports[`services > mirroredModel > importModel > uploads file to S3 on success 1`] = `
{
  "importResult": {
    "newPath": "file/path",
    "sourcePath": "/s3/path/",
  },
  "mirroredModel": {
    "card": {
      "schemaId": "test",
    },
    "settings": {
      "mirror": {
        "destinationModelId": "abc",
      },
    },
  },
}
`;

exports[`services > mirroredModel > importModel > uploads image to S3 on success 1`] = `
{
  "importResult": {
    "image": {
      "imageName": "imageName",
      "imageTag": "tag",
      "modelId": "mirrored-model-id",
    },
  },
  "mirroredModel": {
    "card": {
      "schemaId": "test",
    },
    "settings": {
      "mirror": {
        "destinationModelId": "abc",
      },
    },
  },
}
`;
